azurelinuxagent/common/persist_firewall_rules.py

```diff
@@ -0,0 +1,235 @@
+
+    def __init__(self, dst_ip, uid):
+        """
+        This class deals with ensuring that Firewall rules are persisted over system reboots.
+        It tries to employ using Firewalld.service if present first as it already has provisions for persistent rules.
+        If not, it then creates a new agent-network-setup.service file and copy it over to the osutil.get_systemd_unit_file_install_path() dynamically
+        On top of it, on every service restart it ensures that the WireIP is overwritten and the new IP is blocked as well.
+        """
+        osutil = get_osutil()
+        self._network_setup_service_name = self._AGENT_NETWORK_SETUP_NAME_FORMAT.format(osutil.get_service_name())
+        self._is_systemd = CGroupsApi.is_systemd()
+        self._systemd_file_path = osutil.get_systemd_unit_file_install_path()
+        self._dst_ip = dst_ip
+        self._uid = uid
+        self._wait = osutil.get_firewall_will_wait()
+        # The custom service will try to call the current agent executable to setup the firewall rules
+        self._current_agent_executable_path = os.path.join(os.getcwd(), sys.argv[0])
+
+    @staticmethod
+    def _is_firewall_service_running():
+        # Check if firewall-cmd can connect to the daemon
+        # https://docs.fedoraproject.org/en-US/Fedora/19/html/Security_Guide/sec-Check_if_firewalld_is_running.html
+        # Eg:    firewall-cmd --state
+        #           >   running
+        firewalld_state = PersistFirewallRulesHandler._FIREWALLD_RUNNING_CMD
+        try:
+            return shellutil.run_command(firewalld_state).rstrip() == "running"
+        except Exception as error:
+            logger.info("{0} command returned error/not running: {1}".format(' '.join(firewalld_state), ustr(error)))
+        return False
+
+    def setup(self):
+        if self._is_firewall_service_running():
+            logger.info("Firewalld.service present on the VM, setting up permanent rules on the VM")
+            # In case of a failure, this would throw. In such a case, we don't need to try to setup our custom service
+            # because on system reboot, all iptable rules are reset by firewalld.service so it would be a no-op.
+            self._setup_permanent_firewalld_rules()
+            return
+
+        logger.info(
+            "Firewalld service not running/unavailable, trying to set up {0}".format(self._network_setup_service_name))
+
+        if not CGroupsApi.is_systemd():
+            raise Exception("Systemd not enabled, unable to set {0}".format(self._network_setup_service_name))
+
+        self._setup_network_setup_service()
+
+    def __verify_firewall_rules_enabled(self):
+        # Check if firewall-rules have already been enabled
+        # This function would also return False if the dest-ip is changed. So no need to check separately for that
+        try:
+            AddFirewallRules.check_firewalld_rule_applied(self._dst_ip, self._uid)
```

Comment:
How does this check work? raises both if the rules are not enabled or it if fails for some other reason?

Reply:
If the service is working properly, it would fail if the firewall rules for the specified data (IP, UID) are not applied. 

The check to see if `firewalld.service` is enabled/running on the box is different, its done before this check - `_is_firewall_service_running()`

If the `firewalld.service` is not working properly for external reasons, then it would fail too but that would also mostly fail the initial check, which would make us go the other route and setup our custom service.

I hope that answers your question, if not we can sync offline about this

Reply:
we are logging as verbose, so we won't see the messages under normal runs. seems that we would like to know about some of those, e.g. "If the service is working properly, it would fail if the firewall rules for the specified data (IP, UID) are not applied."

Reply:
In such a case it would still print a generic message like - "Firewall rules not added yet, adding them now using firewalld.service" or "Service: {0} not enabled. Adding it now". 

If you want more data then make it verbose. Initially I added the metadata of these commands as info too (stdout/stderr) but that definitely has a potential of being a bit confusing to the untrained eye so figured I'd make it verbose. Do you have a simpler alternative in mind?